src/Razor/src/Microsoft.VisualStudio.LanguageServerClient.Razor/HtmlCSharp/CompletionHandler.cs

```diff
@@ -108,6 +119,54 @@ internal class CompletionHandler : IRequestHandler<CompletionParams, SumType<Com
             return result;
         }
 
+        internal async Task<CompletionParams> GetProvisionalCompletionParamsAsync(CompletionParams request, LSPDocumentSnapshot documentSnapshot, ProjectionResult projection, CancellationToken cancellationToken)
+        {
+            if (projection.LanguageKind != RazorLanguageKind.Html ||
+                request.Context.TriggerKind != CompletionTriggerKind.TriggerCharacter ||
+                request.Context.TriggerCharacter != ".")
+            {
+                return null;
+            }
+
+            if (projection.Position.Character == 0)
+            {
+                // We're at the start of line. Can't have provisional completions here.
+                return null;
+            }
+
+            var previousCharacterPosition = new Position(projection.Position.Line, projection.Position.Character - 1);
+            var previousCharacterProjection = await _projectionProvider.GetProjectionAsync(documentSnapshot, previousCharacterPosition, cancellationToken).ConfigureAwait(false);
+            if (previousCharacterProjection == null || previousCharacterProjection.LanguageKind != RazorLanguageKind.CSharp)
+            {
+                return null;
+            }
+
+            // Edit the CSharp projected document to contain a '.'. This allows C# completion to provide valid
+            // completion items for moments when a user has typed a '.' that's typically interpreted as Html.
+            if (!(_documentManager is TrackingLSPDocumentManager trackingDocumentManager))
+            {
+                return null;
+            }
+
+            var changes = new[]
+            {
+                new TextChange(TextSpan.FromBounds(previousCharacterProjection.PositionIndex, previousCharacterProjection.PositionIndex), "."),
+            };
+
+            await _joinableTaskFactory.SwitchToMainThreadAsync();
```

Comment:
@NTaylorMullen, we do `SwitchToMainThreadAsync` in a lot of places, but I never see a `SwitchToNonMainThread`, how do we end up ensuring that we don't block the UI thread when we don't need to?

(@ajaybhargavb might be able to answer that question too, I just wanted to be sure Taylor saw it.)

Reply:
`await TaskScheduler.Default;` 😄

Reply:
Ah, and we do do that in a number of spots. Should we do that more widely (IE: always when we're "done" with the main thread) or is the cost of thread switching not worth it?

Reply:
Today we haven't found many situations where the cost is significant enough to warrant it. Definitely something to keep in mind as we're writing the code though